gpu/test/solver/cgs_kernels.cpp

```diff
@@ -131,97 +148,123 @@ class Cgs : public ::testing::Test {
     ASSERT_MTX_NEAR(d_rho, rho, 1e-14);
+    ASSERT_MTX_NEAR(d_alpha, alpha, 1e-14);
+    ASSERT_MTX_NEAR(d_beta, beta, 1e-14);
+    ASSERT_MTX_NEAR(d_gamma, gamma, 1e-14);
 }
 
 
-
 TEST_F(Cgs, GpuCgsStep1IsEquivalentToRef)
 {
     initialize_data();
-    gko::kernels::reference::cgs::step_1(ref, x.get(), r.get(), p.get(),
-                                         q.get(), beta.get(), rho.get());
-    gko::kernels::gpu::cgs::step_1(gpu, d_x.get(), d_r.get(), d_p.get(),
-                                   d_q.get(), d_beta.get(), d_rho.get());
 
-    ASSERT_MTX_NEAR(d_x, x, 1e-14);
-    ASSERT_MTX_NEAR(d_r, r, 1e-14);
+    gko::kernels::reference::cgs::step_1(ref, r.get(), u.get(), p.get(),
+                                         q.get(), beta.get(), rho.get(),
+                                         rho_prev.get());
+    gko::kernels::gpu::cgs::step_1(gpu, d_r.get(), d_u.get(), d_p.get(),
+                                   d_q.get(), d_beta.get(), d_rho.get(),
+                                   d_rho_prev.get());
+
+    ASSERT_MTX_NEAR(d_beta, beta, 1e-14);
+    ASSERT_MTX_NEAR(d_u, u, 1e-14);
     ASSERT_MTX_NEAR(d_p, p, 1e-14);
-    ASSERT_MTX_NEAR(d_q, q, 1e-14);
 }
 
+TEST_F(Cgs, GpuCgsStep2IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::kernels::reference::cgs::step_2(ref, u.get(), v_hat.get(), q.get(),
+                                         t.get(), alpha.get(), rho.get(),
+                                         gamma.get());
+    gko::kernels::gpu::cgs::step_2(gpu, d_u.get(), d_v_hat.get(), d_q.get(),
+                                   d_t.get(), d_alpha.get(), d_rho.get(),
+                                   d_gamma.get());
 
-TEST_F(Cgs, ApplyIsEquivalentToRef)
+    ASSERT_MTX_NEAR(d_alpha, alpha, 1e-14);
+    ASSERT_MTX_NEAR(d_t, t, 1e-14);
+    ASSERT_MTX_NEAR(d_q, q, 1e-14);
+}
+
+TEST_F(Cgs, GpuCgsStep3IsEquivalentToRef)
 {
-    auto mtx = gen_mtx(50, 50);
-    make_spd(mtx.get());
-    auto x = gen_mtx(50, 3);
-    auto b = gen_mtx(50, 3);
-    auto d_mtx = Mtx::create(gpu);
-    d_mtx->copy_from(mtx.get());
-    auto d_x = Mtx::create(gpu);
-    d_x->copy_from(x.get());
-    auto d_b = Mtx::create(gpu);
-    d_b->copy_from(b.get());
-    auto cgs_factory = gko::solver::CgsFactory<>::create(ref, 50, 1e-14);
-    auto d_cgs_factory = gko::solver::CgsFactory<>::create(gpu, 50, 1e-14);
-    auto solver = cgs_factory->generate(std::move(mtx));
-    auto d_solver = d_cgs_factory->generate(std::move(d_mtx));
-
-    solver->apply(b.get(), x.get());
-    d_solver->apply(d_b.get(), d_x.get());
+    initialize_data();
+
+    gko::kernels::reference::cgs::step_3(ref, t.get(), u_hat.get(), r.get(),
+                                         x.get(), alpha.get());
+    gko::kernels::gpu::cgs::step_3(gpu, d_t.get(), d_u_hat.get(), d_r.get(),
+                                   d_x.get(), d_alpha.get());
 
     ASSERT_MTX_NEAR(d_x, x, 1e-14);
+    ASSERT_MTX_NEAR(d_r, r, 1e-14);
 }
-*/
+
 
```

Comment:
I would like to see a test for apply if possible. i.e. similarly to what cg does.

Reply:
I did add a test like suggested.